Add Debug Context Selector

WyoTwT opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CONTRIBUTING.md
-->


#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This PR addresses [Issue 67](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/issues/67).
In a system with multiple children such as the `cdt-gdb-amalgamator`, the Memory Inspector needs a way to select the correct context (as the `cdt-gdb-vscode` Memory Browser does).

This PR has a dependency on `cdt-gdb-amalgamator` [PR 20](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/pull/20) which puts the assignment of the Child IDs in the `cdt-gdb-amalgamator` instead of the Memory Inspector and also uses the `readMemory` since the data returned is in the correct format for Memory Inspector.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Run this branch of the `vscode-memory-inspector` with the mainline `cdt-gdb-vscode` and `cdt-gdb-adapter` along with the modified `cdt-gdb-amalgamator` in PR 20](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/pull/20).

Start the debugger as a Desktop Extension. In the newly launched VSCode, open the cdt-amalgamator/sampleWorkspace folder and place a breakpoint at the first instruction in empty1.c's main and empty2.c's main.  Start debugging with `Amalgamator Example` and you should run to the breakpoints. Open a new Memory Inspector with the Command Palette's `Memory: Show Memory Inspector` to open a new view. The dropdown should list both `proc1` and `proc2` listed in the Call Stack.



#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the contribution guidelines](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CONTRIBUTING.md)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the code of conduct](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/main/CODE_OF_CONDUCT.md)

colin-grant-work commented on src/common/debug-requests.ts:
```diff
@@ -18,13 +18,30 @@
+}
+export interface ChildrenResponse extends DebugProtocol.Response {
+    body: ChildContexts;
+}
```
Let's move this to `amalgamator-gdb-tracker.ts`, along with the `getChildContexts` request, since it's not DAP.

colin-grant-work commented on src/common/debug-requests.ts:
```diff
@@ -18,13 +18,30 @@
+export interface Context {
+    name: string;
+    id: number;
+}
```
Let's move this one to `messaging.ts`.

colin-grant-work commented on src/common/messaging.ts:
```diff
@@ -38,6 +38,8 @@ export type StoreMemoryResult = void;
 export type ApplyMemoryArguments = URI | undefined;
 export type ApplyMemoryResult = MemoryOptions;
 
+export type GetChildContextsResult = DebugRequestTypes['getChildContexts'][1];
```
I'm still getting hung up on the word 'child' here. Maybe just `getContexts` or `getAvailableContexts`.

WyoTwT replied:
This was moved to amalgamator-gdb-tracker.ts (since it is only used there) and renamed to GetContextsResult.

colin-grant-work commented on src/common/messaging.ts:
```diff
@@ -55,15 +57,17 @@ export const sessionContextChangedType: NotificationType<SessionContext> = { met
-export const writeMemoryType: RequestType<WriteMemoryArguments, WriteMemoryResult> = { method: 'writeMemory' };
+export const readMemoryType: RequestType<[ReadMemoryArguments, Context?], ReadMemoryResult> = { method: 'readMemory' };
+export const writeMemoryType: RequestType<[WriteMemoryArguments, Context?], WriteMemoryResult> = { method: 'writeMemory' };
 export const getVariablesType: RequestType<ReadMemoryArguments, VariableRange[]> = { method: 'getVariables' };
 export const storeMemoryType: RequestType<StoreMemoryArguments, void> = { method: 'storeMemory' };
 export const applyMemoryType: RequestType<ApplyMemoryArguments, ApplyMemoryResult> = { method: 'applyMemory' };
```
Let's add the possibility of a `Context` to these as well, even if we won't be using it yet.

colin-grant-work commented on src/plugin/memory-provider.ts:
```diff
@@ -143,24 +143,34 @@ export class MemoryProvider {
+        const session = this.assertCapability('supportsReadMemoryRequest', 'read memory');
+        const handler = this.adapterRegistry?.getHandlerForSession(session.type);
+        if (handler?.readMemory) { return handler.readMemory(session, args[0], args[1]); }
+        return sendRequest(session, 'readMemory', args[0]);
```
Destructuring might make this clearer:
In `messaging.ts`

```typescript
type ReadMemoryWithContext = [ReadMemoryArguments, Context?];
```

```suggestion
    public async readMemory([readMemoryArgs, context]: ReadMemoryWithContext): Promise<ReadMemoryResult> {
        const session = this.assertCapability('supportsReadMemoryRequest', 'read memory');
        const handler = this.adapterRegistry?.getHandlerForSession(session.type);
        if (handler?.readMemory) { return handler.readMemory(session, readMemoryArgs, context); }
        return sendRequest(session, 'readMemory', readMemoryArgs);
```

colin-grant-work commented on src/webview/components/memory-widget.tsx:
```diff
@@ -102,6 +106,9 @@ export class MemoryWidget extends React.Component<MemoryWidgetProps, MemoryWidge
                 applyMemory={this.props.applyMemory}
+                childContexts={this.props.childContexts}
+                child={this.props.child}
+                setChild={this.props.setChild}
```
Same here and in the rest of the frontend code re: my hangup about the term 'child'.

WyoTwT commented on src/plugin/memory-storage.ts:
I don't feel comfortable with the undefined in the readMemory (line 73 below) but it felt like I was changing too many functions which is what these more surgical changes were supposed to avoid (instead of adding Context to readArguments).

WyoTwT committed:
Move ReadMemory/WriteMemory into adapter-capabilities

Since the ReadMemory/WriteMemory may be dependent upon the DAPs, move
into the adapter-capabilities.ts file.

Signed-off-by: Thor Thayer <thor.thayer@ericsson.com>

WyoTwT committed:
Adding Context for cdt-amalgamator and other DAPs

Populate the Context Dropdown from the different debug Contexts
if available so different DAPs can be addressed.
Add an optional Context to queries so the query can be directed
to the appropriate handler.
Add support for cdt-amalgamator.

Signed-off-by: Thor Thayer <thor.thayer@ericsson.com>

WyoTwT committed:
Match Current Context with one of Contexts

Match current context id with one of the elements of the
Contexts array.

Signed-off-by: Thor Thayer <thor.thayer@ericsson.com>

WyoTwT committed:
Conditional 'Show In Memory Inspector' for Variable

Add new conditional for context menu showing Memory Inspector for a
variable because some DAPs may not support it.

Signed-off-by: Thor Thayer <thor.thayer@ericsson.com>

colin-grant-work commented on src/plugin/adapter-registry/adapter-capabilities.ts:
```diff
@@ -122,15 +128,24 @@ export class AdapterVariableTracker implements vscode.DebugAdapterTracker {
-    getSizeOfVariable?(variableName: string, session: vscode.DebugSession): Promise<bigint | undefined>;
+    getSizeOfVariable?(variableName: string, session: vscode.DebugSession, context?: Context): Promise<bigint | undefined>;
+
+    readMemory?(session: vscode.DebugSession, params: ReadMemoryArguments, context: Context): Promise<ReadMemoryResult>;
+
+    writeMemory?(session: vscode.DebugSession, params: WriteMemoryArguments, context: Context): Promise<WriteMemoryResult>;
```
In other cases, the context is optional; why is it required here?

WyoTwT replied:
Whoops. Good catch. Thanks!

colin-grant-work commented on src/webview/components/options-widget.tsx:
```diff
@@ -127,6 +131,30 @@ export class OptionsWidget extends React.Component<OptionsWidgetProps, OptionsWi
+
+    protected showContexts = () => {
+        if (this.props.contexts.length === 0) {
+            return false;
```
`undefined` would be more natural here.

WyoTwT commented:
Since this is a local draft PR, I'm closing it.

WyoTwT commented:
Since this is a local draft PR, I'm closing it. See upstream PR:
https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/133
